src/mds/MDSPinger.cc

```diff
@@ -0,0 +1,109 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "common/dout.h"
+
+#include "mds/MDSRank.h"
+#include "mds/MDSPinger.h"
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_mds
+#undef dout_prefix
+#define dout_prefix *_dout << "mds.pinger " << __func__
+
+MDSPinger::MDSPinger(MDSRank *mds)
+  : mds(mds) {
+}
+
+void MDSPinger::send_ping(mds_rank_t rank) {
+  dout(10) << ": rank=" << rank << dendl;
+
+  std::scoped_lock locker(lock);
+
+  auto it = ping_state_by_rank.find(rank);
+  if (it == ping_state_by_rank.end()) {
+    dout(20) << ": init ping pong state for rank=" << rank << dendl;
+    it = ping_state_by_rank.emplace(rank, PingState()).first;
+    it->second.last_acked_time = clock::now();
+  }
```

Comment:
this code can be simplified:

````
auto [it, inserted] = ping_state_by_rank.emplace(rank, {});
if (inserted) {
   dout(20) << ": init ping pong state for rank=" << rank << dendl;
}
````

(you already initialize `last_acked_time` in the header)